src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs

```diff
@@ -38,27 +38,26 @@ internal static partial class WebSocketValidate
             SearchValues.Create("!#$%&'*+-.0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ^_`abcdefghijklmnopqrstuvwxyz|~");
 
         internal static void ThrowIfInvalidState(WebSocketState currentState, bool isDisposed, WebSocketState[] validStates)
+            => ThrowIfInvalidState(currentState, isDisposed, innerException: null, validStates ?? []);
+
+        internal static void ThrowIfInvalidState(WebSocketState currentState, bool isDisposed, Exception? innerException, WebSocketState[]? validStates = null)
         {
-            string validStatesText = string.Empty;
+            if (validStates is not null && Array.IndexOf(validStates, currentState) == -1)
+            {
+                string invalidStateMessage = SR.Format(
+                    SR.net_WebSockets_InvalidState, currentState, string.Join(", ", validStates));
+
+                throw new WebSocketException(WebSocketError.InvalidState, invalidStateMessage, innerException);
+            }
 
-            if (validStates != null && validStates.Length > 0)
+            if (innerException is not null)
             {
-                foreach (WebSocketState validState in validStates)
-                {
-                    if (currentState == validState)
-                    {
-                        // Ordering is important to maintain .NET 4.5 WebSocket implementation exception behavior.
-                        ObjectDisposedException.ThrowIf(isDisposed, typeof(WebSocket));
-                        return;
-                    }
-                }
-
-                validStatesText = string.Join(", ", validStates);
+                Debug.Assert(currentState == WebSocketState.Aborted);
+                throw new OperationCanceledException(nameof(WebSocketState.Aborted), innerException);
             }
 
-            throw new WebSocketException(
-                WebSocketError.InvalidState,
-                SR.Format(SR.net_WebSockets_InvalidState, currentState, validStatesText));
+            // Ordering is important to maintain .NET 4.5 WebSocket implementation exception behavior.
```

Comment:
Didn't we change the ordering though?

Reply:
No, the order is still the same (InvalidStateEx -> ODE).

If we don't count the added keep-alive OCEs (for now), then

<details>
<summary>the old code</summary>

https://github.com/dotnet/runtime/blob/e047cce05b54a77b668edc1c53cfb10d36b325c5/src/libraries/Common/src/System/Net/WebSockets/WebSocketValidate.cs#L40-L62

</details>

has the following logic:
1. Traverse valid states collection
2. If we traversed all the valid states and our state is not found
    * throw InvalidStateEx
3. If our state is found
    * check for disposed
        * throw ODE if needed

which gives you ordering InvalidStateEx -> ODE (if the state is invalid, we don't check disposed flag)

Here the exceptions have the same ordering (checking the status first, disposed second), but without (a bit) confusing foreach (that I failed to see any real value in, as even the string with the list of valid states was combined later in a separate string.Join)

Reply:
If we bring the keep-alive OCEs into picture, we need to look at Receive in general; for this mental exercise it's easier think of keep-alive exceptions as if it was a user triggered abort:

The old code was
1. Checking the state only _before_ trying to hold the mutex (ThrowIfInvalidState with valid read states)
    - If abort already happened: InvalidStateEx (wrapped into OCE) in both old and new code
2. Inside the mutex, if any exception occurred:
    - (2.1) If it was OCE -- and it could only be caused by a user CT -- we leave all the OCEs as-is
    - (2.2) If it was not OCE, BUT the state is Aborted -- we decide it's caused by (either user's or CT callback triggered) Abort -- and wrap InvalidStateEx into OCE, and the "real" exception is inner
    - (2.3) everything else is WebSocket exception
  
The change brings the (2.2) check a bit earlier execution-wise, but the order remains the same (because OCEs are just left as-is, and ours is already "prepared"). This is done to make sure the "real" keep-alive exception propagates before the ODE (because ODE was _caused_ by the Abort).

in the old code, that "real" Abort exception was not exactly interesting (because it was always an ODE, if I'm not mistaken -- the fact of Abort itself was the most interesting thing -- and the user should be already aware of Abort happening, bc they either aborted the WebSocket, or cancelled the CT on their side)

But with the keep-alives Abort can happen "from inside", so now we need to be able to tell the difference.